platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/ValueNodeMap.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * Tracks the values of inputs and selects on the server and client side.
+ */
+class ValueNodeMap {
+	private readonly states = new Map<HTMLElement, ValueNodeState>();
+
+	/**
+	 * Adds the given value node to this map.
+	 */
+	public addNode(node: HTMLElement) {
+		this.states.set(node, new ValueNodeState(node));
+	}
+
+	/**
+	 * Approves all pending values.
+	 */
+	public approveNodeValues() {
+		this.states.forEach((state, node) => state.approveValue());
+	}
+
+	/**
+	 * Transfers all changed values of value nodes to the specified data structure.
+	 */
+	public copyNodeValues(data: any) {
+		this.states.forEach((state, node) => state.updatePendingValue(data));
+	}
+
+	public isValueChanged(node: HTMLElement) {
+		return this.getState(node).isValueChanged();
+	}
+
+	public setValue(node: HTMLInputElement, value: string) {
+		node.value = value;
+		this.getState(node).assumeValue(value);
+
+		// FIXME this ad-hoc call looks like a hack
+		AUTO_COMPLETE_ENGINE.setCommittedValue(node, value);
+	}
+
+	public setSelectedOptions(select: HTMLSelectElement, options: HTMLOptionElement[]) {
+		let type = select.type;
+		if(type == "select-one") {
+			options[0].selected = true;
+		} else if(type == "select-multiple") {
+			for(let option of select.options) {
+				option.selected = options.includes(option);
+			}
+		} else {
+			alert("Unknown select type " + type + ".");
```

Comment:
- Punctuation seems odd... should be `"Unknown select type: " + type` or so.
- We should probably add the ID of the offending node to the error message.
- Why would we use a native alert in the first place? (rather than logging to console, or using `throw new Error("...")` much like a couple of lines below)